Shared/Server Protocol/CodableError.swift

```diff
@@ -0,0 +1,241 @@
+            switch self
+            {
+            case .unknown: return nil
+            case .string(let string): return string
+            case .number(let number): return number
+            case .error(let error): return error
+            case .codableError(let error): return error.error
+            case .array(let array): return array
+            case .dictionary(let dictionary): return dictionary
+            }
+        }
+        
+        init(_ rawValue: Any?)
+        {
+            switch rawValue
+            {
+            case let string as String: self = .string(string)
+            case let number as Int: self = .number(number)
+            case let error as NSError: self = .codableError(CodableError(error: error))
+            case let array as [Any]: self = .array(array.compactMap(UserInfoValue.init))
+            case let dictionary as [String: Any]: self = .dictionary(dictionary.compactMapValues(UserInfoValue.init))
+            default: self = .unknown
+            }
+        }
+        
+        init(from decoder: Decoder) throws
+        {
+            let container = try decoder.singleValueContainer()
+
+            if
+                let data = try? container.decode(Data.self),
+                let error = try? NSKeyedUnarchiver.unarchivedObject(ofClass: NSError.self, from: data)
+            {
+                self = .error(error)
+            }
+            else if let codableError = try? container.decode(CodableError.self)
+            {
+                self = .codableError(codableError)
+            }
+            else if let string = try? container.decode(String.self)
+            {
+                self = .string(string)
+            }
+            else if let number = try? container.decode(Int.self)
+            {
+                self = .number(number)
+            }
+            else if let array = try? container.decode([UserInfoValue].self)
+            {
+                self = .array(array)
+            }
+            else if let dictionary = try? container.decode([String: UserInfoValue].self)
+            {
+                self = .dictionary(dictionary)
+            }
+            else
+            {
+                self = .unknown
+            }
+        }
+        
+        func encode(to encoder: Encoder) throws
+        {
+            var container = encoder.singleValueContainer()
+            
+            let codableValue: Codable?
+            
+            switch self
+            {
+            case .codableError(let codableError): codableValue = codableError
+            case .error(let error):
+                // Sanitize error to ensure it can be encoded correctly.
+                let sanitizedError = error.sanitizedForSerialization()
+                let data = try NSKeyedArchiver.archivedData(withRootObject: sanitizedError, requiringSecureCoding: true)
+                codableValue = data
+                
+            default: codableValue = self.value as? Codable
```

Comment:
Make sure self.value is a Codable representation, not the wrapped value (relevant for arrays and dictionaries)